packages/pds/src/app-view/subscription/repo.ts

```diff
@@ -0,0 +1,187 @@
+          for await (const msg of sub) {
+            const ops = await getOps(msg)
+            await db.transaction(async (tx) => {
+              await this.handleOps(tx, ops, msg.time)
+              await this.setState(tx, { cursor: msg.seq })
+            })
+          }
+        })
+        if (ran && !this.destroyed) {
+          throw new Error('Repo sub completed, but should be persistent')
+        }
+      } catch (err) {
+        appViewLogger.error(
+          { err, service: this.service },
+          'repo subscription errored',
+        )
+      }
+      if (!this.destroyed) {
+        await wait(5000 + jitter(1000))
+      }
+    }
+  }
+
+  destroy() {
+    this.destroyed = true
+    this.leader.destroy(new DisconnectError())
+  }
+
+  private async handleOps(
+    tx: Database,
+    ops: PreparedWrite[],
+    timestamp: string,
+  ) {
+    const { services } = this.ctx
+    const indexingTx = services.appView.indexing(tx)
+    for (const op of ops) {
+      if (
+        op.action === WriteOpAction.Create ||
+        op.action === WriteOpAction.Update
+      ) {
+        await indexingTx.indexRecord(op.uri, op.cid, op.record, timestamp)
+      } else if (op.action === WriteOpAction.Delete) {
+        await indexingTx.deleteRecord(op.uri)
+      } else {
+        const exhaustiveCheck: never = op
+        throw new Error(`Unsupported op: ${exhaustiveCheck?.['action']}`)
+      }
+    }
+  }
+
+  async getState(): Promise<State> {
+    const sub = await this.ctx.db.db
+      .selectFrom('subscription')
+      .selectAll()
+      .where('service', '=', this.service)
+      .where('method', '=', METHOD)
+      .executeTakeFirst()
+    return sub ? (JSON.parse(sub.state) as State) : { cursor: -1 } // @TODO change to 0 once fix on pds lands
+  }
+
+  private async setState(tx: Database, state: State): Promise<void> {
+    tx.assertTransaction()
+    const res = await tx.db
+      .updateTable('subscription')
+      .where('service', '=', this.service)
+      .where('method', '=', METHOD)
+      .set({ state: JSON.stringify(state) })
+      .executeTakeFirst()
+    if (res.numUpdatedRows < 1) {
+      await tx.db
+        .insertInto('subscription')
+        .values({
+          service: this.service,
+          method: METHOD,
+          state: JSON.stringify(state),
+        })
+        .executeTakeFirst()
+    }
+  }
+
+  private getSubscription(opts: { signal: AbortSignal }) {
+    return new Subscription({
+      service: this.service,
+      method: METHOD,
+      signal: opts.signal,
+      getParams: () => this.getState(),
+      onReconnectError: (err, reconnects, initial) => {
+        appViewLogger.warn(
+          { err, reconnects, initial },
+          'repo subscription reconnect',
+        )
+      },
+      validate: (value) => {
+        try {
+          return lexicons.assertValidXrpcMessage<Message>(METHOD, value)
+        } catch (err) {
+          appViewLogger.warn(
+            { err, service: this.service },
```

Comment:
maybe worth logging the message or some parts of it if possible?

this may help with debugging BGS if it sends bad messages 🤔